### PR TITLE
fix(#1639): ensure tree autocommands match filetype as well as name

### LINKED ISSUE #1639: Strict Matching Of NvimTree Buffers
#1629 demonstated incorrect handling of buffers named `NvimTree*` on startup.

There are further cases, mostly autocommands, in which similar tests will need to be applied. Something like:

```lua
  if opts.view.centralize_selection then
    create_nvim_tree_autocmd("BufEnter", {
      pattern = "NvimTree_*",
      callback = function()
        if utils.is_current_buf_nvim_tree() then
          vim.schedule(function()
            local keys = api.nvim_replace_termcodes("zz", true, false, true)
            api.nvim_feedkeys(keys, "n", true)
          end)
        end
      end,
    })
  end
```

It's probably safe to just test the filetype as it should be set at most of these points.

### PR DESCRIPTION
**Pull request by alex-courtis**
fixes #1639 

requires #1629 

requires #1632

requires #1637

Autocommands are firing for buffers named `*NvimTree_*` which are real files.

### CODE REVIEW

**Comment by alex-courtis** [comment]
Test cases:
- [ ] hijack_cursor
- [ ] wipe tree, no quit_on_open
- [ ] enter tree, no watchers and reload_on_bufenter
- [ ] leave window, float.enable and float.quit_on_focus_loss
- [ ] leave window with filter, float.enable and float.quit_on_focus_loss
- [ ] buf leave, full_name
- [ ] cursor move, full_name

File: lua/nvim-tree.lua [hunk 1]
```diff
@@ -370,8 +377,16 @@ local function setup_autocommands(opts)
   end
 
   if not opts.actions.open_file.quit_on_open then
-    create_nvim_tree_autocmd("BufWipeout", { pattern = "NvimTree_*", callback = view._prevent_buffer_override })
+    create_nvim_tree_autocmd("BufWipeout", {
+      pattern = "NvimTree_*",
+      callback = function()
+        if api.nvim_buf_get_option(0, "filetype") == "NvimTree" then
```
Comment (kyazdani42) [inline lua/nvim-tree.lua:383]: vim.bo.filetype is more luaish

Comment (alex-courtis) [reply lua/nvim-tree.lua:383]: Today I Learned

Comment (gegoune) [reply lua/nvim-tree.lua:383]: `vim.bo.filetype` is also a shorthand for `vim.bo[0].filetype`.

Comment (alex-courtis) [reply lua/nvim-tree.lua:383]: Fantastic. I might merge these checks into `is_nvim_tree_buf`.

**Review by kyazdani42 — APPROVED** [review]

**Pull request merged**